pandas_gbq/gbq.py

```diff
@@ -577,141 +500,97 @@ def run_query(self, query, **kwargs):
                 raise AccessDenied(
                     "The credentials have been revoked or expired, "
                     "please re-run the application to re-authorize")
-        except HttpError as ex:
+        except self.http_error as ex:
             self.process_http_error(ex)
 
-        job_reference = query_reply['jobReference']
-        job_id = job_reference['jobId']
+        job_id = query_reply.job_id
         self._print('Job ID: %s\nQuery running...' % job_id)
 
-        while not query_reply.get('jobComplete', False):
+        while query_reply.state != 'DONE':
             self.print_elapsed_seconds('  Elapsed', 's. Waiting...')
 
             timeout_ms = job_config['query'].get('timeoutMs')
             if timeout_ms and timeout_ms < self.get_elapsed_seconds() * 1000:
                 raise QueryTimeout('Query timeout: {} ms'.format(timeout_ms))
 
+            timeout_sec = 1.0
+            if timeout_ms:
+                # Wait at most 1 second so we can show progress bar
+                timeout_sec = min(1.0, timeout_ms / 1000.0)
+
             try:
-                query_reply = job_collection.getQueryResults(
-                    projectId=job_reference['projectId'],
-                    jobId=job_id).execute()
-            except HttpError as ex:
+                query_reply.result(timeout=timeout_sec)
+            except TimeoutError:
+                # Use our own timeout logic
+                pass
+            except self.http_error as ex:
                 self.process_http_error(ex)
 
         if self.verbose:
-            if query_reply['cacheHit']:
+            if query_reply.cache_hit:
                 self._print('Query done.\nCache hit.\n')
             else:
-                bytes_processed = int(query_reply.get(
-                    'totalBytesProcessed', '0'))
-                self._print('Query done.\nProcessed: {}'.format(
-                    self.sizeof_fmt(bytes_processed)))
+                bytes_processed = query_reply.total_bytes_processed or 0
+                bytes_billed = query_reply.total_bytes_billed or 0
+                self._print('Query done.\nProcessed: {} Billed: {}'.format(
+                    self.sizeof_fmt(bytes_processed),
+                    self.sizeof_fmt(bytes_billed)))
                 self._print('Standard price: ${:,.2f} USD\n'.format(
-                    bytes_processed * self.query_price_for_TB))
+                    bytes_billed * self.query_price_for_TB))
 
             self._print('Retrieving results...')
```

Comment:
Presumably this is never going to be hit because the prior part is blocking?

Reply:
Yes and no. I think it indeed is less relevant, but actually fetching the rows and constructing the dataframe will take non-zero time, especially for larger result sets.